/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists.length==0) return null;
        PriorityQueue<ListNode> pq=new PriorityQueue<>((a,b)-> a.val-b.val);
        for(ListNode list:lists){
            if(list!=null) pq.offer(list);
        }
        ListNode head=new ListNode(0);
        ListNode res=head;
        while(!pq.isEmpty()){
            ListNode temp=pq.poll();
            if(temp.next!=null) pq.offer(temp.next);
            res.next=temp;
            res=res.next;
        }
        return head.next;
    }
}
